# Examining Racial Discrimination in the US Job Market

### Background
Racial discrimination continues to be pervasive in cultures throughout the world. Researchers examined the level of racial discrimination in the United States labor market by randomly assigning identical résumés to black-sounding or white-sounding names and observing the impact on requests for interviews from employers.

### Data
In the dataset provided, each row represents a resume. The 'race' column has two values, 'b' and 'w', indicating black-sounding and white-sounding. The column 'call' has two values, 1 and 0, indicating whether the resume received a call from employers or not.

Note that the 'b' and 'w' values in race are assigned randomly to the resumes when presented to the employer.

### Exercises
You will perform a statistical analysis to establish whether race has a significant impact on the rate of callbacks for resumes.

Answer the following questions **in this notebook below and submit to your Github account**. 

   1. What test is appropriate for this problem? Does CLT apply?
   2. What are the null and alternate hypotheses?
   3. Compute margin of error, confidence interval, and p-value. Try using both the bootstrapping and the frequentist statistical approaches.
   4. Write a story describing the statistical significance in the context or the original problem.
   5. Does your analysis mean that race/name is the most important factor in callback success? Why or why not? If not, how would you amend your analysis?

You can include written notes in notebook cells using Markdown: 
   - In the control panel at the top, choose Cell > Cell Type > Markdown
   - Markdown syntax: http://nestacms.com/docs/creating-content/markdown-cheat-sheet

#### Resources
+ Experiment information and data source: http://www.povertyactionlab.org/evaluation/discrimination-job-market-united-states
+ Scipy statistical methods: http://docs.scipy.org/doc/scipy/reference/stats.html 
+ Markdown syntax: http://nestacms.com/docs/creating-content/markdown-cheat-sheet
+ Formulas for the Bernoulli distribution: https://en.wikipedia.org/wiki/Bernoulli_distribution

In [1]:
import pandas as pd
import numpy as np
from scipy import stats

In [3]:
data = pd.io.stata.read_stata('D:/DataScienceFoundation/SpringBoard/miniProject/EDA_racial_discrimination/data/us_job_market_discrimination.dta')

In [9]:
# number of callbacks for white-sounding names
sum(data[data.race=='w'].call)

235.0

In [8]:
# number of callbacks for black-sounding names
sum(data[data.race=='b'].call)

157.0

In [12]:
len(data)

4870

In [5]:
data.head()

id ad  education  ofjobs  yearsexp  honors  volunteer  military  empholes  \
0  b  1          4       2         6       0          0         0         1   
1  b  1          3       3         6       0          1         1         0   
2  b  1          4       1         6       0          0         0         0   
3  b  1          3       4         6       0          1         0         1   
4  b  1          3       3        22       0          0         0         0   

   occupspecific    ...      compreq  orgreq  manuf  transcom  bankreal trade  \
0             17    ...          1.0     0.0    1.0       0.0       0.0   0.0   
1            316    ...          1.0     0.0    1.0       0.0       0.0   0.0   
2             19    ...          1.0     0.0    1.0       0.0       0.0   0.0   
3            313    ...          1.0     0.0    1.0       0.0       0.0   0.0   
4            313    ...          1.0     1.0    0.0       0.0       0.0   0.0   

  busservice othservice  missind  ownership  
0        0.0        0.0      0.0             
1        0.0        0.0      0.0             
2        0.0        0.0      0.0             
3        0.0        0.0      0.0             
4        0.0        1.0      0.0  Nonprofit  

[5 rows x 65 columns]

## 1. What test is appropriate for this problem? Does CLT apply?

#### Centeral Limit Theorem stats that the mean of the sample shoould resamble a normal distribution or if the large number of samples is generated. Becuase the sample is large, n=4870 then we can say CLT applies. 

#### A two sample t-test is appropriate since we are comparing two group means.

## 2. What are the null and alternate hypotheses?

* H0: There is no difference between black callback rate and white callback rate 
* Ha: There is a significant different between black callback rate and white callback rate 

## 3. Compute margin of error, confidence interval, and p-value. Try using both the bootstrapping and the frequentist statistical approaches.

#### Frequentist Approach

In [16]:
black_count = len(data[data.race=='b'])
black_call_sum = sum(data[data.race=='b'].call)
white_count = len(data[data.race=='w'])
white_call_sum =sum(data[data.race=='w'].call)

In [53]:

mean_call_b = np.mean(data[data.race=='b'].call)
mean_call_w = np.mean(data[data.race=='w'].call)
diff_of_means = mean_call_w - mean_call_b

std_call = np.std(data.call)
std_call_b = np.std(data[data.race=='b'].call)
std_call_w = np.std(data[data.race=='w'].call)

print('The callback rate for resumes with black names is: ', mean_call_b)
print('The callback rate for resumes with white names is: ', mean_call_w)

The callback rate for resumes with black names is:  0.0644763857126236
The callback rate for resumes with white names is:  0.09650924056768417


In [50]:
# Hypothesis Testing via Frequentist Approach
se = np.sqrt(std_call**2/black_count + std_call**2/white_count)
z = (diff_of_means-0)/se
p = 2*stats.norm.sf(z)
print('Standard Error (SE) = %.3f' %se)
print('Difference of Means Observed (w-b) = %.3f' %(diff_of_means))
print('Z-statistic for the observed mean difference is %.3f' %z) 
print('p-value with z-statistics = %.3e' %p)


Standard Error (SE) = 0.008
Difference of Means Observed (w-b) = 0.032
Z-statistic for the observed mean difference is 4.108
p-value with z-statistics = 3.983e-05



#### Margin of Error and Confidence Interval of the observed value
#### Use 5% significance level of the observed value for MOE and CI calculation


In [54]:
se = np.sqrt(std_call_b**2/black_count + std_call_w**2/white_count)
margin_of_error = stats.norm.ppf(.975)*se
confidence_interval = [diff_of_means-margin_of_error, diff_of_means+margin_of_error]
print('Difference of Means Observed (white and black) = %.3f' %(diff_of_means))
print('Margin of Error  for the Frequentist Method = %.3f' %margin_of_error)
print('95%% Confidence interval of Observed Mean Difference (Frequentist Method) = [%.3f, %.3f]' %(confidence_interval[0], confidence_interval[1]))

Difference of Means Observed (w-b) = 0.032
Margin of Error (Frequentist Method) = 0.015
95% Confidence interval of Observed Mean Difference (Frequentist Method) = [0.017, 0.047]


#### Bootstrap Method

In [58]:

size = 100000
bs_replicates = np.empty(size)
diff_of_means = mean_call_w - mean_call_b
all_calls = np.concatenate((data[data.race=='w'].call, data[data.race=='b'].call))
for i in range(size):
    bs_samples = np.random.permutation(all_calls)
    bs_samples_w = bs_samples[:len(data[data.race=='w'].call)]
    bs_samples_b = bs_samples[len(data[data.race=='w'].call):]
    bs_replicates[i] = np.mean(bs_samples_w) - np.mean(bs_samples_b)
p = np.sum(np.abs(bs_replicates)>np.abs(diff_of_means))/len(bs_replicates)
print('Difference of Means Observed (w-b) = %.3f' %(diff_of_means))
print('p-value for the observed diff of means = %.1e' %p)

Difference of Means Observed (w-b) = 0.032
p-value for the observed diff of means = 1.0e-05


In [61]:
size = 10000
bs_replicates = np.empty(size)
diff_of_means = mean_call_w - mean_call_b
for i in range(size):
    bs_samples_w = np.random.choice(data[data.race=='w'].call, len(data[data.race=='w'].call))
    bs_samples_b = np.random.choice(data[data.race=='b'].call, len(data[data.race=='b'].call))
    bs_replicates[i] = np.mean(bs_samples_w) - np.mean(bs_samples_b)

confidence_interval = np.percentile(bs_replicates, [0.05, 99.95])

print('99.9%% Confidence interval is = [%.3f, %.3f]' %(confidence_interval[0], confidence_interval[1]))

99.9% Confidence interval is = [0.008, 0.057]


## 4. Write a story describing the statistical significance in the context or the original problem.

#### Resumes with black-sounding names receive a callback rate of 6.4%, but white names receive a callback rate of 9.7%.  Moreover, the 99% confience interval suggests that the true callback rate difference could range from 1.2 percentage points to 5.2 percentage points. Therefore, racial discrmination in the U.S. labor market still appears to be a continual challenge.

## 5. Does your analysis mean that race/name is the most important factor in callback success? Why or why not? If not, how would you amend your analysis?

####  Although the above experiment shows race is important factor for hiring, we cannot conclude on race as most important factor since there are very other differnt factors like skills, education, work history. Therefore, we cannot conclude that race/name is the most important factor